File: app/src/main/java/awais/instagrabber/webservices/DirectMessagesService.java
package awais.instagrabber.webservices;

import androidx.annotation.NonNull;

import com.google.common.collect.ImmutableMap;

import org.json.JSONArray;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import awais.instagrabber.repositories.DirectMessagesRepository;
import awais.instagrabber.repositories.requests.directmessages.BroadcastOptions;
import awais.instagrabber.repositories.requests.directmessages.BroadcastOptions.ThreadIdOrUserIds;
import awais.instagrabber.repositories.requests.directmessages.LinkBroadcastOptions;
import awais.instagrabber.repositories.requests.directmessages.PhotoBroadcastOptions;
import awais.instagrabber.repositories.requests.directmessages.TextBroadcastOptions;
import awais.instagrabber.repositories.requests.directmessages.VideoBroadcastOptions;
import awais.instagrabber.repositories.requests.directmessages.VoiceBroadcastOptions;
import awais.instagrabber.repositories.responses.directmessages.DirectBadgeCount;
import awais.instagrabber.repositories.responses.directmessages.DirectInboxResponse;
import awais.instagrabber.repositories.responses.directmessages.DirectThreadBroadcastResponse;
import awais.instagrabber.repositories.responses.directmessages.DirectThreadFeedResponse;
import awais.instagrabber.utils.TextUtils;
import awais.instagrabber.utils.Utils;
import retrofit2.Call;
import retrofit2.Retrofit;

public class DirectMessagesService extends BaseService {
    private static final String TAG = "DiscoverService";

    private static DirectMessagesService instance;

    private final DirectMessagesRepository repository;
    private final String csrfToken;
    private final String userId;
    private final String deviceUuid;

    private DirectMessagesService(@NonNull final String csrfToken,
                                  @NonNull final String userId,
                                  @NonNull final String deviceUuid) {
        this.csrfToken = csrfToken;
        this.userId = userId;
        this.deviceUuid = deviceUuid;
        final Retrofit retrofit = getRetrofitBuilder()
                .baseUrl("https://i.instagram.com")
                .build();
        repository = retrofit.create(DirectMessagesRepository.class);
    }

    public String getCsrfToken() {
        return csrfToken;
    }

    public String getUserId() {
        return userId;
    }

    public String getDeviceUuid() {
        return deviceUuid;
    }

    public static DirectMessagesService getInstance(@NonNull final String csrfToken,
                                                    @NonNull final String userId,
                                                    @NonNull final String deviceUuid) {
        if (instance == null
                || !Objects.equals(instance.getCsrfToken(), csrfToken)
                || !Objects.equals(instance.getUserId(), userId)
                || !Objects.equals(instance.getDeviceUuid(), deviceUuid)) {
            instance = new DirectMessagesService(csrfToken, userId, deviceUuid);
        }
        return instance;
    }

    public Call<DirectInboxResponse> fetchInbox(final String cursor,
                                                final long seqId) {
        final ImmutableMap.Builder<String, Object> queryMapBuilder = ImmutableMap.<String, Object>builder()
                .put("visual_message_return_type", "unseen")
                .put("thread_message_limit", 10)
                .put("persistentBadging", true)
                .put("limit", 10);
        if (!TextUtils.isEmpty(cursor)) {
            queryMapBuilder.put("cursor", cursor);
            queryMapBuilder.put("direction", "older");
        }
        if (seqId != 0) {
            queryMapBuilder.put("seq_id", seqId);
        }
        return repository.fetchInbox(queryMapBuilder.build());
    }

    public Call<DirectThreadFeedResponse> fetchThread(final String threadId,
                                                      final String cursor) {
        final ImmutableMap.Builder<String, Object> queryMapBuilder = ImmutableMap.<String, Object>builder()
                .put("visual_message_return_type", "unseen")
                .put("limit", 10)
                .put("direction", "older");
        if (!TextUtils.isEmpty(cursor)) {
            queryMapBuilder.put("cursor", cursor);
        }
        return repository.fetchThread(threadId, queryMapBuilder.build());
    }

    public Call<DirectBadgeCount> fetchUnseenCount() {
        return repository.fetchUnseenCount();
    }

    public Call<DirectThreadBroadcastResponse> broadcastText(final String clientContext,
                                                             final ThreadIdOrUserIds threadIdOrUserIds,
                                                             final String text) {
        final List<String> urls = TextUtils.extractUrls(text);
        if (!urls.isEmpty()) {
            return broadcastLink(clientContext, threadIdOrUserIds, text, urls);
        }
        return broadcast(new TextBroadcastOptions(clientContext, threadIdOrUserIds, text));
    }

    public Call<DirectThreadBroadcastResponse> broadcastLink(final String clientContext,
                                                             final ThreadIdOrUserIds threadIdOrUserIds,
                                                             final String linkText,
                                                             final List<String> urls) {
        return broadcast(new LinkBroadcastOptions(clientContext, threadIdOrUserIds, linkText, urls));
    }

    public Call<DirectThreadBroadcastResponse> broadcastPhoto(final String clientContext,
                                                              final ThreadIdOrUserIds threadIdOrUserIds,
                                                              final String uploadId) {
        return broadcast(new PhotoBroadcastOptions(clientContext, threadIdOrUserIds, true, uploadId));
    }

    public Call<DirectThreadBroadcastResponse> broadcastVideo(final String clientContext,
                                                              final ThreadIdOrUserIds threadIdOrUserIds,
                                                              final String uploadId,
                                                              final String videoResult,
                                                              final boolean sampled) {
        return broadcast(new VideoBroadcastOptions(clientContext, threadIdOrUserIds, videoResult, uploadId, sampled));
    }

    public Call<DirectThreadBroadcastResponse> broadcastVoice(final String clientContext,
                                                              final ThreadIdOrUserIds threadIdOrUserIds,
                                                              final String uploadId,
                                                              final List<Float> waveform,
                                                              final int samplingFreq) {
        return broadcast(new VoiceBroadcastOptions(clientContext, threadIdOrUserIds, uploadId, waveform, samplingFreq));
    }

    private Call<DirectThreadBroadcastResponse> broadcast(@NonNull final BroadcastOptions broadcastOptions) {
        if (TextUtils.isEmpty(broadcastOptions.getClientContext())) {
            throw new IllegalArgumentException("Broadcast requires a valid client context value");
        }
        final Map<String, Object> form = new HashMap<>();
        if (!TextUtils.isEmpty(broadcastOptions.getThreadId())) {
            form.put("thread_id", broadcastOptions.getThreadId());
        } else {
            form.put("recipient_users", new JSONArray(broadcastOptions.getUserIds()).toString());
        }
        form.put("_csrftoken", csrfToken);
        form.put("_uid", userId);
        form.put("__uuid", deviceUuid);
        form.put("client_context", broadcastOptions.getClientContext());
        form.put("mutation_token", broadcastOptions.getClientContext());
        form.putAll(broadcastOptions.getFormMap());
        form.put("action", "send_item");
        final Map<String, String> signedForm = Utils.sign(form);
        return repository.broadcast(broadcastOptions.getItemType().getValue(), signedForm);
    }
}

File: app/src/main/java/awais/instagrabber/webservices/DirectMessagesService.kt
package awais.instagrabber.webservices

import awais.instagrabber.repositories.DirectMessagesRepository
import awais.instagrabber.repositories.requests.directmessages.*
import awais.instagrabber.repositories.responses.directmessages.*
import awais.instagrabber.repositories.responses.giphy.GiphyGif
import awais.instagrabber.utils.TextUtils.extractUrls
import awais.instagrabber.utils.Utils
import org.json.JSONArray
import java.util.*

object DirectMessagesService : BaseService() {
    private val repository: DirectMessagesRepository = RetrofitFactory.retrofit.create(DirectMessagesRepository::class.java)

    suspend fun fetchInbox(
        cursor: String?,
        seqId: Long,
    ): DirectInboxResponse {
        val queryMap = mutableMapOf(
            "visual_message_return_type" to "unseen",
            "thread_message_limit" to 10.toString(),
            "persistentBadging" to true.toString(),
            "limit" to 10.toString(),
        )
        if (!cursor.isNullOrBlank()) {
            queryMap["cursor"] = cursor
            queryMap["direction"] = "older"
        }
        if (seqId != 0L) {
            queryMap["seq_id"] = seqId.toString()
        }
        return repository.fetchInbox(queryMap)
    }

    suspend fun fetchThread(
        threadId: String,
        cursor: String?,
    ): DirectThreadFeedResponse {
        val queryMap = mutableMapOf(
            "visual_message_return_type" to "unseen",
            "limit" to 20.toString(),
            "direction" to "older",
        )
        if (!cursor.isNullOrBlank()) {
            queryMap["cursor"] = cursor
        }
        return repository.fetchThread(threadId, queryMap)
    }

    suspend fun fetchUnseenCount(): DirectBadgeCount = repository.fetchUnseenCount()

    suspend fun broadcastText(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        text: String,
        repliedToItemId: String?,
        repliedToClientContext: String?,
    ): DirectThreadBroadcastResponse {
        val urls = extractUrls(text)
        if (urls.isNotEmpty()) {
            return broadcastLink(csrfToken, userId, deviceUuid, clientContext, threadIdOrUserIds, text, urls, repliedToItemId, repliedToClientContext)
        }
        val broadcastOptions = TextBroadcastOptions(clientContext, threadIdOrUserIds, text)
        if (!repliedToItemId.isNullOrBlank() && !repliedToClientContext.isNullOrBlank()) {
            broadcastOptions.repliedToItemId = repliedToItemId
            broadcastOptions.repliedToClientContext = repliedToClientContext
        }
        return broadcast(csrfToken, userId, deviceUuid, broadcastOptions)
    }

    private suspend fun broadcastLink(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        linkText: String,
        urls: List<String>,
        repliedToItemId: String?,
        repliedToClientContext: String?,
    ): DirectThreadBroadcastResponse {
        val broadcastOptions = LinkBroadcastOptions(clientContext, threadIdOrUserIds, linkText, urls)
        if (!repliedToItemId.isNullOrBlank() && !repliedToClientContext.isNullOrBlank()) {
            broadcastOptions.repliedToItemId = repliedToItemId
            broadcastOptions.repliedToClientContext = repliedToClientContext
        }
        return broadcast(csrfToken, userId, deviceUuid, broadcastOptions)
    }

    suspend fun broadcastPhoto(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        uploadId: String,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, PhotoBroadcastOptions(clientContext, threadIdOrUserIds, true, uploadId))

    suspend fun broadcastVideo(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        uploadId: String,
        videoResult: String,
        sampled: Boolean,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, VideoBroadcastOptions(clientContext, threadIdOrUserIds, videoResult, uploadId, sampled))

    suspend fun broadcastVoice(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        uploadId: String,
        waveform: List<Float>,
        samplingFreq: Int,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, VoiceBroadcastOptions(clientContext, threadIdOrUserIds, uploadId, waveform, samplingFreq))

    suspend fun broadcastStoryReply(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        text: String,
        mediaId: String,
        reelId: String,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, StoryReplyBroadcastOptions(UUID.randomUUID().toString(), threadIdOrUserIds, text, mediaId, reelId))

    suspend fun broadcastReaction(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        itemId: String,
        emoji: String?,
        delete: Boolean,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, ReactionBroadcastOptions(clientContext, threadIdOrUserIds, itemId, emoji, delete))

    suspend fun broadcastAnimatedMedia(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        giphyGif: GiphyGif,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, AnimatedMediaBroadcastOptions(clientContext, threadIdOrUserIds, giphyGif))

    suspend fun broadcastMediaShare(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        clientContext: String,
        threadIdOrUserIds: ThreadIdOrUserIds,
        mediaId: String,
    ): DirectThreadBroadcastResponse =
        broadcast(csrfToken, userId, deviceUuid, MediaShareBroadcastOptions(clientContext, threadIdOrUserIds, mediaId))

    private suspend fun broadcast(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        broadcastOptions: BroadcastOptions,
    ): DirectThreadBroadcastResponse {
        require(broadcastOptions.clientContext.isNotBlank()) { "Broadcast requires a valid client context value" }
        val form = mutableMapOf<String, Any>(
            "_csrftoken" to csrfToken,
            "_uid" to userId,
            "__uuid" to deviceUuid,
            "client_context" to broadcastOptions.clientContext,
            "mutation_token" to broadcastOptions.clientContext,
        )
        val threadId = broadcastOptions.threadId
        if (!threadId.isNullOrBlank()) {
            form["thread_id"] = threadId
        } else {
            val userIds = broadcastOptions.userIds
            require(!userIds.isNullOrEmpty()) {
                "Either provide a thread id or pass a list of user ids"
            }
            form["recipient_users"] = JSONArray(userIds).toString()
        }
        val repliedToItemId = broadcastOptions.repliedToItemId
        val repliedToClientContext = broadcastOptions.repliedToClientContext
        if (!repliedToItemId.isNullOrBlank() && !repliedToClientContext.isNullOrBlank()) {
            form["replied_to_item_id"] = repliedToItemId
            form["replied_to_client_context"] = repliedToClientContext
        }
        form.putAll(broadcastOptions.formMap)
        form["action"] = "send_item"
        val signedForm = Utils.sign(form)
        return repository.broadcast(broadcastOptions.itemType.value, signedForm)
    }

    suspend fun addUsers(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        userIds: Collection<Long>,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "user_ids" to JSONArray(userIds).toString(),
        )
        return repository.addUsers(threadId, form)
    }

    suspend fun removeUsers(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        userIds: Collection<Long>,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "user_ids" to JSONArray(userIds).toString(),
        )
        return repository.removeUsers(threadId, form)
    }

    suspend fun updateTitle(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        title: String,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "title" to title,
        )
        return repository.updateTitle(threadId, form)
    }

    suspend fun addAdmins(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        userIds: Collection<Long>,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "user_ids" to JSONArray(userIds).toString(),
        )
        return repository.addAdmins(threadId, form)
    }

    suspend fun removeAdmins(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        userIds: Collection<Long>,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "user_ids" to JSONArray(userIds).toString(),
        )
        return repository.removeAdmins(threadId, form)
    }

    suspend fun deleteItem(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        itemId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.deleteItem(threadId, itemId, form)
    }

    suspend fun rankedRecipients(
        mode: String?,
        showThreads: Boolean?,
        query: String?,
    ): RankedRecipientsResponse {
        // String correctedMode = mode;
        // if (TextUtils.isEmpty(mode) || (!mode.equals("raven") && !mode.equals("reshare"))) {
        //     correctedMode = "raven";
        // }
        val queryMap = mutableMapOf<String, String>()
        if (!mode.isNullOrBlank()) {
            queryMap["mode"] = mode
        }
        if (!query.isNullOrBlank()) {
            queryMap["query"] = query
        }
        if (showThreads != null) {
            queryMap["showThreads"] = showThreads.toString()
        }
        return repository.rankedRecipients(queryMap)
    }

    suspend fun forward(
        toThreadId: String,
        itemType: String,
        fromThreadId: String,
        itemId: String,
    ): DirectThreadBroadcastResponse {
        val form = mapOf(
            "action" to "forward_item",
            "thread_id" to toThreadId,
            "item_type" to itemType,
            "forwarded_from_thread_id" to fromThreadId,
            "forwarded_from_thread_item_id" to itemId,
        )
        return repository.forward(form)
    }

    suspend fun createThread(
        csrfToken: String,
        userId: Long,
        deviceUuid: String,
        userIds: List<Long>,
        threadTitle: String?,
    ): DirectThread {
        val userIdStringList = userIds.map { it.toString() }
        val form = mutableMapOf<String, Any>(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "_uid" to userId,
            "recipient_users" to JSONArray(userIdStringList).toString(),
        )
        if (!threadTitle.isNullOrBlank()) {
            form["thread_title"] = threadTitle
        }
        val signedForm = Utils.sign(form)
        return repository.createThread(signedForm)
    }

    suspend fun mute(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid
        )
        return repository.mute(threadId, form)
    }

    suspend fun unmute(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.unmute(threadId, form)
    }

    suspend fun muteMentions(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.muteMentions(threadId, form)
    }

    suspend fun unmuteMentions(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.unmuteMentions(threadId, form)
    }

    suspend fun participantRequests(
        threadId: String,
        pageSize: Int,
        cursor: String? = null,
    ): DirectThreadParticipantRequestsResponse {
        return repository.participantRequests(threadId, pageSize, cursor)
    }

    suspend fun approveParticipantRequests(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        userIds: List<Long>,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "user_ids" to JSONArray(userIds).toString(),
            // "share_join_chat_story" to String.valueOf(true)
        )
        return repository.approveParticipantRequests(threadId, form)
    }

    suspend fun declineParticipantRequests(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        userIds: List<Long>,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "user_ids" to JSONArray(userIds).toString(),
        )
        return repository.declineParticipantRequests(threadId, form)
    }

    suspend fun approvalRequired(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.approvalRequired(threadId, form)
    }

    suspend fun approvalNotRequired(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.approvalNotRequired(threadId, form)
    }

    suspend fun leave(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.leave(threadId, form)
    }

    suspend fun end(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): DirectThreadDetailsChangeResponse {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.end(threadId, form)
    }

    suspend fun fetchPendingInbox(cursor: String?, seqId: Long): DirectInboxResponse {
        val queryMap = mutableMapOf(
            "visual_message_return_type" to "unseen",
            "thread_message_limit" to 20.toString(),
            "persistentBadging" to true.toString(),
            "limit" to 10.toString(),
        )
        if (!cursor.isNullOrBlank()) {
            queryMap["cursor"] = cursor
            queryMap["direction"] = "older"
        }
        if (seqId != 0L) {
            queryMap["seq_id"] = seqId.toString()
        }
        return repository.fetchPendingInbox(queryMap)
    }

    suspend fun approveRequest(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.approveRequest(threadId, form)
    }

    suspend fun declineRequest(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
    ): String {
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
        )
        return repository.declineRequest(threadId, form)
    }

    suspend fun markAsSeen(
        csrfToken: String,
        deviceUuid: String,
        threadId: String,
        directItem: DirectItem,
    ): DirectItemSeenResponse? {
        val itemId = directItem.itemId ?: return null
        val form = mapOf(
            "_csrftoken" to csrfToken,
            "_uuid" to deviceUuid,
            "use_unified_inbox" to "true",
            "action" to "mark_seen",
            "thread_id" to threadId,
            "item_id" to itemId,
        )
        return repository.markItemSeen(threadId, itemId, form)
    }
}